File: graph/bst/bst.py
class Node:
    def __init__(self, val):
        self.left = None
        self.right = None
        self.val = val


class BST:
    def __init__(self, val):
        self.root = None
    
    def add_node(self, val, node):
        if not node:
            return Node(val)
        if val < node.val:
            node.left = self.add_node(val, node.left)
        else:
            node.right = self.add_node(val, node.right)
        return node
    
    def add(self, val):
        self.root = self.add_node(val, self.root)

    def del_max(self):  # does not work if root is max
        v, prev = self, self
        while v.right:
            prev = v
            v = v.right
        prev.right = None
        return v.val
    
    def rm(self, val):
        def aux(node):
            if node.val == val:
                if not node.left:
                    return node.right
                if self.left.right:  # can call self.left.del_max
                    self.val = self.left.del_max()
                else:
                    self = self.left
            elif val < self.val:
                self.left = self.left.rm(val)
            else:
                self.right = self.right.rm(val)
            return self

    def __contains__(self, val):
        def aux(node):
            if not node:
                return False
            if node.val == val:
                return True
            if val < node.val:
                return aux(node.left)
            else:
                return aux(node.right)
        return aux(self.root)

File: graph/bst/bst.cpp

struct bst {
    bst* left;
    bst* right;
    int val;
};

bst* create(int v) {
    bst* b = new bst();
    b->val = v;
    return b;
}

void add(bst* b, int v) {  // b must be non-empty
    if(v < b->val)
        if(!(b->left))
            b->left = create(v);
        else
            add(b->left, v);
    else if(!(b->right))
            b->right = create(v);
        else
            add(b->right, v);
}

void rm(bst* b, int v) {
    if(v == b->val) {
        bst* b1 = b->left;
        if(!b1)
            *b = *(b->right);
        else {
            while(b1->right)
                b1 = b1->right;
            b1->right = b->right;
            *b = *b1;
        }
    }
    else if(v < b->val) 
        rm(b->left, v);
    else
        rm(b->right, v);
}

int main() {
    bst* b = create(6);
    add(b, 8);
    add(b, 2);
    add(b, 5);
    add(b, 7);
    add(b, 9);
    rm(b, 8);
    rm(b, 6);
    rm(b, 9);
    rm(b, 5);
}